### PR TITLE
ddl: Speed up add index phase

### PR DESCRIPTION
**Pull request by zimulala**
Handle issue #2257 
- Use `BatchGet` instead of `Get`

- Remove the judgment of `Exist`

Using 3 TiKV, 1 PD, 1 TiDB , the servers of TiKV and PD on one computer, the server of TiDB on another computer.
The lease of TiDB is 1s.
The number of records in this table is 336253, and the table's structure is as follows:
+------------------+------------------+------+------+---------+-------+
| Field            | Type             | Null | Key  | Default | Extra |
+------------------+------------------+------+------+---------+-------+
| a           | varchar(64)      | YES  |      |         |       |
| b           | bigint(20)       | NO   |      | 0       |       |
| c           | datetime         | YES  |      | NULL    |       |
| d           | int(11) UNSIGNED | NO   |      | 0       |       |
| e           | int(11) UNSIGNED | NO   |      | 0       |       |
| f            | int(11) UNSIGNED | NO   |      | 0       |       |
| g           | varchar(28)      | NO   |      |         |       |
| h           | varchar(28)      | NO   |      |         |       |
+------------------+------------------+------+------+---------+-------+

tidb-before> alter table t add index id (b);
Query OK, 0 rows affected (6 min 2.55 sec)
tidb-before> alter table t add index ids (e);
Query OK, 0 rows affected (6 min 9.32 sec)

tidb-current> alter table t add index id (b);
Query OK, 0 rows affected (1 min 6.78 sec)
tidb-current> mysql> alter table t add index ids (e);
Query OK, 0 rows affected (1 min 3.15 sec)

### CODE REVIEW

**Comment by zimulala** [comment]
PATL @coocood @tiancaiamao

File: ddl/index.go [hunk 1]
```diff
@@ -325,30 +325,56 @@ func (d *ddl) onDropIndex(t *meta.Meta, job *model.Job) error {
...
+	[]*indexRecord, error) {
+	// Through handles access to get all row keys and column maps.
 	cols := t.Cols()
-	colMap := make(map[int64]*types.FieldType)
-	for _, v := range indexInfo.Columns {
-		col := cols[v.Offset]
-		colMap[col.ID] = &col.FieldType
+	handlesLen := len(handles)
+	rowKeys := make([]kv.Key, 0, handlesLen)
+	colMaps := make([]map[int64]*types.FieldType, 0, handlesLen)
```
Comment (shenli) [inline ddl/index.go]: Should we use a single colMaps for all handles?

File: ddl/index.go [hunk 2]
```diff
@@ -432,41 +458,44 @@ func (d *ddl) getSnapshotRows(t table.Table, version uint64, seekHandle int64) (
...
+		err = txn.LockKeys(idxRecord.key)
 		if err != nil {
 			return 0, errors.Trace(err)
 		}
 
 		// Create the index.
-		_, err = kvIdx.Create(txn, vals, handle)
+		handle, err := kvIdx.Create(txn, idxRecord.vals, idxRecord.handle)
 		if err != nil {
+			if terror.ErrorEqual(err, kv.ErrKeyExists) && idxRecord.handle == handle {
```
Comment (shenli) [inline ddl/index.go]: Will we get this error here?

File: ddl/index.go (hunk 1 above)
Comment (zimulala) [inline ddl/index.go]: This  colMaps is for all handles.

**Comment by zimulala** [comment]
PTAL @shenli

File: ddl/index.go [hunk 3]
```diff
@@ -325,30 +325,54 @@ func (d *ddl) onDropIndex(t *meta.Meta, job *model.Job) error {
...
+func (d *ddl) fetchRowColVals(txn kv.Transaction, t table.Table, handles []int64, indexInfo *model.IndexInfo) (
+	[]*indexRecord, error) {
+	// Through handles access to get all row keys.
+	handlesLen := len(handles)
+	rowKeys := make([]kv.Key, 0, handlesLen)
+	idxRecords := make([]*indexRecord, handlesLen)
+	for i, h := range handles {
+		rowKey := tablecodec.EncodeRecordKey(t.RecordPrefix(), h)
+		rowKeys = append(rowKeys, rowKey)
+		idxRecords[i] = &indexRecord{handle: h, key: rowKey}
```
Comment (hanfei1991) [inline ddl/index.go]: why not append it ?

File: ddl/index.go [hunk 4]
```diff
@@ -325,30 +325,54 @@ func (d *ddl) onDropIndex(t *meta.Meta, job *model.Job) error {
...
 	for _, v := range indexInfo.Columns {
 		col := cols[v.Offset]
-		vals = append(vals, row[col.ID])
+		colMap[col.ID] = &col.FieldType
+	}
+	for _, idxRecord := range idxRecords {
+		rowVal, ok := pairMap[string(idxRecord.key)]
+		if !ok {
+			// Row doesn't exist, skip it.
+			idxRecord.notExist = true
```
Comment (hanfei1991) [inline ddl/index.go]: When does the idxRecord not exist ?

File: ddl/index.go [hunk 5]
```diff
@@ -325,30 +325,54 @@ func (d *ddl) onDropIndex(t *meta.Meta, job *model.Job) error {
...
+	pairMap, err := snap.BatchGet(rowKeys)
 	if err != nil {
-		return nil, nil, errors.Trace(err)
+		return nil, errors.Trace(err)
 	}
-	vals := make([]types.Datum, 0, len(indexInfo.Columns))
+
+	// Get corresponding values for pairMap.
+	cols := t.Cols()
+	colMap := make(map[int64]*types.FieldType)
```
Comment (hanfei1991) [inline ddl/index.go]: Everytime we need construct a colMap. Can we store it ?

File: ddl/index.go (hunk 4 above)
Comment (zimulala) [inline ddl/index.go]: This row be deleted.

File: ddl/index.go (hunk 5 above)
Comment (zimulala) [inline ddl/index.go]: Now don't need. It be built for every 1024 rows.

File: ddl/index.go (hunk 3 above)
Comment (zimulala) [inline ddl/index.go]: Both are OK. It's more simple.

Comment (zimulala) [inline ddl/index.go]: Line 336 only appends an object already exist, line 337 needs to append a new object.

File: ddl/index.go (hunk 4 above)
Comment (hanfei1991) [inline ddl/index.go]: Then the comment can be "Row has been deleted, so we skip it"?

**Comment by disksing** [comment]
It will be better to break `fetchRowColVals` into following steps:
1) Encode `handles []uint64` to `rowkeys []kv.Key`.
2) `BatchGet(rowKeys)`, you will get `pairMap map[kv.Key][]byte`.
3) Iterate `rowKeys` to build `idxRecords []*indexRecord`, just ignore keys that are absent, you don't  need the `notExist` flag.

**Comment by zimulala** [comment]
PTAL @disksing @hanfei1991 @shenli @coocood

**Comment by hanfei1991** [comment]
lgtm

File: ddl/index.go [hunk 6]
```diff
@@ -432,41 +456,39 @@ func (d *ddl) getSnapshotRows(t table.Table, version uint64, seekHandle int64) (
...
-		exist, _, err := kvIdx.Exist(txn, vals, handle)
-		if err != nil {
-			return 0, errors.Trace(err)
-		} else if exist {
-			// Index already exists, skip it.
-			nextHandle = handle
-			continue
-		}
-		err = txn.LockKeys(rowKey)
+	nextHandle := handles[0]
```
Comment (disksing) [inline ddl/index.go]: Why do we need this variable? Look like we can just return `idxRecords[len(idxRecords)-1].handle` at last line.

**Comment by coocood** [comment]
LGTM

**Pull request merged**